src/librustc_mir/shim.rs

```diff
@@ -294,7 +294,7 @@ fn build_clone_shim<'a, 'tcx>(tcx: TyCtxt<'a, 'tcx, 'tcx>,
 {
     debug!("build_clone_shim(def_id={:?})", def_id);
 
-    let mut builder = CloneShimBuilder::new(tcx, def_id);
+    let mut builder = CloneShimBuilder::new(tcx, def_id, self_ty);
```

Comment:
Without this commit everything ICEs. Everything.

Any code instantiating a clone shim of tuples or arrays or closures  (libcore does) will break because the type used will be TySelf (not TyArray or TyClosure or whatever) and we'll get errors because we're trying to call `.index()` or `.field()` on a place that can't be indexed. I can leave a comment to this effect somewhere.

Reply:
I mean I can see why this is wrong, but I'm a bit confused -- the code works *now*, afaik, right? Or is this just patching up some earlier commit of yours?

Reply:
It breaks once the change to the array shim is made - anything that depends on the type (e.g. indexing or enum discriminants) starts breaking because it'd see an unsubstituted `Self`.

Reply:
Thsi commit should be earlier in the series I think.

We basically don't hit it because we write to temporary Places and then copy and that works well. I don't recall the exact details, but eddyb can explain.

Reply:
ok good enough

Reply:
Oh right yeah this doesn't break the tuple stuff but it does break the array stuff (and the enum stuff from my other PR)